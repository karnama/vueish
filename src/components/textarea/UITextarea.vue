<template>
    <div class="ui-text">
        <label :for="$attrs.id ?? name" class="text-sm text-gray-500">
            {{ label }}
        </label>

        <div class="group relative container">
            <div class="flex overflow-hidden relative">
                <textarea :id="$attrs.id ?? name"
                          v-bind="$attrs"
                          ref="input"
                          v-model="value"
                          :name="name"
                          :disabled="disabled"
                          class="flex-1 appearance-none bg-transparent transition-border-color leading-tight
                                 focus:outline-none rounded-none transition-text-color disabled:cursor-not-allowed
                                 disabled:text-gray-400"
                          :style="disabled || fixed ? 'resize: none' : ''" />

                <span v-if="disabled"
                      class="h-5 w-5 text-gray-400"
                      v-html="lockIcon" />


                <span v-else-if="!noClear && modelValue"
                      class="clear-icon h-5 w-5 cursor-pointer right-0 top-1 group-hover:opacity-100 transition-opacity
                            text-gray-500 relative -mr-5 group-hover:mr-0 -mt-1 transition-spacing"
                      @click="value = ''"
                      v-html="clearIcon" />
            </div>

            <div class="absolute w-full border-b left-0 bottom-0" />
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue';
import { SetupReturn } from '@/types';
import {
    autofocus,
    label,
    autofocusElement,
    noClear,
    name,
    disabled, useVModel
} from '@composables/input';
import { getIcon } from '@/helpers';

export default defineComponent({
    name: 'UIText',

    inheritAttrs: false,

    props: {
        modelValue: {
            type: String,
            required: true
        },

        fixed: Boolean,
        label,
        autofocus,
        noClear,
        name,
        disabled
    },

    emits: ['update:modelValue'],

    setup(props): SetupReturn {
        const input = ref<HTMLInputElement>();
        const value = useVModel<string>(props);
        const lockIcon = getIcon('lock');
        const clearIcon = getIcon('clear');
        autofocusElement(props.autofocus, input);

        return {
            value,
            input,
            lockIcon,
            clearIcon
        };
    }
});
</script>
