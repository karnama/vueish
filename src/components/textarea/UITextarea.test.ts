import { mount } from '@vue/test-utils';
import UITextarea from './UITextarea.vue';

describe('UITextarea', () => {
    it('should handle model-binding correctly', async () => {
        const wrapper = mount({
            template: '<div><UITextarea v-model="input" name="input"/></div>',
            components: { UITextarea },
            props: {},
            data() {
                return { input: '' };
            }
        });

        await wrapper.get('textarea').setValue('Hello World');

        expect(wrapper.emitted()).toHaveProperty('update:modelValue');
        expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual(['Hello World']);
    });

    it('should be enabled by default', () => {
        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name: 'input'
            }
        });

        expect(wrapper.get('textarea').attributes().disabled).toBeUndefined();
    });

    it('should be disabled when the prop is provided', async () => {
        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name: 'input',
                disabled: false
            }
        });

        const input = wrapper.get('textarea');

        expect(input.attributes().disabled).toBeUndefined();
        await wrapper.setProps({ disabled: true });
        expect(input.attributes().disabled).not.toBeUndefined();
    });

    it('should prevent resize when the fixed prop is given', async () => {
        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name: 'input',
                fixed: false
            }
        });

        const input = wrapper.get('textarea');

        expect(input.element.style.resize).toBe('');
        await wrapper.setProps({ fixed: true });
        expect(input.element.style.resize).toBe('none');
    });

    it('should correctly assign the name prop', () => {
        const name = 'input';

        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name
            }
        });

        const input = wrapper.get('textarea');
        const label = wrapper.get('label');

        expect(input.attributes().id).toBe(name);
        expect(input.attributes().name).toBe(name);
        expect(label.attributes().for).toBe(name);
    });

    it('should correctly display the label when passed as a prop', () => {
        const label = 'text';

        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name: 'input',
                label
            }
        });

        expect(wrapper.get('label').text()).toBe(label);
    });

    it('should autofocus on the input when the prop is provided', () => {
        const elem = document.createElement('div');
        document.body.appendChild(elem);

        const wrapper = mount(UITextarea, {
            props: {
                modelValue: '',
                name: 'input',
                autofocus: true
            },
            attachTo: elem
        });

        expect(wrapper.get('textarea').element).toBe(document.activeElement);
    });

    it('should not show the clear icon if no-clear is provided', () => {
        const wrapper = mount(UITextarea, {
            props: {
                name: 'input',
                modelValue: 'Hello World',
                noClear: true
            }
        });

        expect(wrapper.find('.clear-icon').exists()).toBe(false);
    });

    it('should not show the clear icon if there is no value', () => {
        const wrapper = mount(UITextarea, {
            props: {
                name: 'input',
                modelValue: ''
            }
        });

        expect(wrapper.find('.clear-icon').exists()).toBe(false);
    });

    it('should clear the value when the clear icon is clicked', async () => {
        const modelValue = 'Hello World';

        const wrapper = mount(UITextarea, {
            props: {
                name: 'input',
                modelValue
            }
        });

        const input = wrapper.get('textarea').element;

        expect(input.value).toBe(modelValue);
        await wrapper.get('.clear-icon').trigger('click');
        expect(wrapper.emitted()).toHaveProperty('update:modelValue');
        expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual(['']);
    });
});
