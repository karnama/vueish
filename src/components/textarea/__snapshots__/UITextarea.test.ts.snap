// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UITextarea should display correctlt 1`] = `
<div
  class="ui-textarea relative"
>
  <label
    class="font-medium text-gray-700"
    for="input"
  />
  <div
    class="group relative shadow border border-gray-300 rounded bg-white
                    transition focus-within:border-blue-400"
  >
    <div
      class="flex items-center items-stretch"
    >
      <textarea
        class="flex-1 p-3 appearance-none bg-transparent outline-none
                                 disabled:cursor-not-allowed disabled:text-gray-400"
        id="input"
        name="input"
      />
      <div
        class="flex flex-col justify-between"
      >
        <!--v-if-->
      </div>
    </div>
  </div>
  <!--v-if-->
</div>
`;
